### DISCUSSION TITLE
LDM Notes for June 10th, 2024

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#8211] [LDM Notes]
https://github.com/dotnet/csharplang/blob/main/meetings/2024/LDM-2024-06-10.md

### Agenda

- `ref struct`s implementing interfaces and in generics

This agenda comes with an explicit request for you; yes, you reading this discussion! Do you have a use case for `ref struct`s implementing interfaces? If so, we want to hear about it, so we can ensure that we're building the right feature. Please let us hear your use cases!

**Comment by colejohnson66** [comment]
Off the bat: ref structs implementing interfaces removes the need for duck typing, like `foreach` and `await` use. Future language features could use interfaces to define the structure.

**Reply by 333fred** [reply]
Not particularly. Again, `Span<T>` _cannot_ implement `IEnumerable`, and we cannot base the codegen of `foreach` purely on `IEnumerable`, as `IEnumerator` is too costly. Without that, the entire idea of throwing away structural typing here doesn't hold water.

**Reply by cap5lut** [reply]
for me came the `IDisposable` duck typing to mind, tho that alone would not be worth it.

**Reply by neon-sunset** [reply]
For that to work effectively, C# would probably need an associated type support, in order to allow the inference of the enumerator type, which itself would have to satisfy the `allows ref struct`, err, anti-constraint. This could enable Span<T> and other types subject to ref lifetimes to be supported by (yet another) struct LINQ implementation, though this time around being much closer to what is offered by Rust.

**Reply by colejohnson66** [reply]
> Not particularly. Again, `Span<T>` _cannot_ implement `IEnumerable`, and we cannot base the codegen of `foreach` purely on `IEnumerable`, as `IEnumerator` is too costly. Without that, the entire idea of throwing away structural typing here doesn't hold water.

I’m not suggesting requiring it for existing features, as that would break backwards compatibility. I’m only talking about future ones.

**Reply by 333fred** [reply]
> I’m only talking about future ones.

And we couldn't do that for `foreach` if we were redesigning it from the ground up _today_ with `IEnumerable<T>`. While I'm also for `IEnumerable2` that uses associated types to get better enumerator generics to flow through the system, we don't have that, and we won't have that for C# 13. What we're looking for here is **concrete** use cases that we can validate the design of `ref struct`s implementing interfaces against. Not hypotheticals that wouldn't work without still other language features.

**Reply by Richiban** [reply]
I think the C# compiler is (arguably) embracing duck-typing more and more, rather than less. Patterns essentially compile down into method / property / indexer calls, and there might be multiple of them in a single pattern.

**Comment by agocke** [comment]
FYI if you're willing to operate on Enumerators directly, you can still get a lot of the benefit in .NET 9:

```C#
using System;
using System.Collections;
using System.Collections.Generic;

List<int> list = [ 1, 2, 3];
Span<int> span = [ 1, 2, 3];

C.M(list.GetEnumerator());
C.M(span.GetEnumerator2());

public class C {
    public static void M<TEnum>(TEnum enumerator)
        where TEnum : IEnumerator<int>, allows ref struct
    {
        while (enumerator.MoveNext())
        {
            Console.WriteLine(enumerator.Current);
        }
    }
}

// Stand-in code until we implement IEnumerator for span enumerator
static class SpanExt
{
    public static SpanEnum<T> GetEnumerator2<T>(this Span<T> span)
        => new SpanEnum<T>(span.GetEnumerator());
}

ref struct SpanEnum<T> : IEnumerator<T>
{    
    private readonly Span<T>.Enumerator _enum;
    public SpanEnum(Span<T>.Enumerator e)
    {
        _enum = e;
    }
    
    
    public T Current => _enum.Current;

    object IEnumerator.Current => _enum.Current!;

    public void Dispose()
    {
    }

    public bool MoveNext()
    {
        return _enum.MoveNext();
    }

    public void Reset()
    {
        throw new NotImplementedException();
    }
}
```

**Reply by yaakov-h** [reply]
Looking over it I was expecting to maybe be able to write `M<TEnumerable, TEnumerator>` but that won't work either without a new interface `IEnumerable<TElement, TEnumerator> where TEnumerator : TEnumerator<TElement>` or similar.

**Comment by yaakov-h** [comment]
One use case I've come across but never really needed that much is a way to pass a Span around that I can also be finished with early. I expect with C# 13 I could do something roughly like:

```cs
interface ITemporarySpannable<T> : IDisposable
{
    public Span<T> Span { get; }
}

explicit extension SpannableMemory<T> for Memory<T> : ITemporarySpannable<T>
{
    // Memory<T> already has a Span property, not sure if we would need to redeclare it here?
    public void Dispose() { } // no-op as this has no cleanup
}

explicit extension SpannableSelf<T> for Span<T> : ITemporarySpannable<T>
{
    public Span<T> Span => return this;
    public void Dispose() { } // no-op as this has no cleanup
}

ref struct ArrayLease<T> : ITemporarySpannable<T>
{
    public ArrayLease(ArrayPool<T> pool, int size)
    {
        this.pool = pool;
        this.array = pool.Rent(size);
    }

    T[] array;
    ArrayPool<T> pool;
    
    public Span<T> Span => array.AsSpan();

    public void Dispose() => pool.Return(array); // return to pool when we are finished, even if still in scope.

    // missing IsDisposed checks etc.
}
```

This might be a dumb idea but I do keep finding myself wishing that ArrayPool was a little bit easier to work with than var/try/finally every time.

**Reply by 333fred** [reply]
Well, you were never going to be able to implement interfaces with extensions in C# 13; that was, at best, going to be a preview, and almost certainly won't make the cut at all. But the last example is somewhat interesting.

**Comment by jjonescz** [comment]
With `ref struct` interfaces, the `lock` statement could be extended to allow other types than just the new `Lock` - e.g., `SpinLock`. See the first bullet point in https://github.com/dotnet/csharplang/blob/3ab828dcfe2ae5eafcc8537e71ac543eb2927340/proposals/lock-object.md#alternatives.

**Comment by Ollhax** [comment]
Here's my a use case for ref structs with interfaces: I'm working on an entity component system where I have my data in one struct `ComponentData` and a facade struct `Component`:

```
public struct ComponentData
{
	public float UpdateTime;
}

public readonly struct Component(FarmSystem system, ComponentIndex index) : IComponentView<Component>
{
	public static Component Create(FarmSystem system, ComponentIndex index) => new(system, index);

	public readonly ref ComponentData Data => ref System.GetData(Index);
	public readonly FarmSystem System = system;
	public readonly ComponentIndex Index = index;

	public readonly float UpdateTime { get => Data.UpdateTime; set => System.SetUpdateTime(value); }
}
```

What I'd like to do is to make Component a `ref struct` and store the `ComponentData` reference in the struct, so I can avoid the cost of accessing it each time use the `Data` property. I still need the `IComponentView<Component>` interface to provide the static `Create` factory function to get the machinery to work.

**Comment by hez2010** [comment]
> One of the marquee original intentions for ref structs implementing interfaces was for Span<T> and ReadOnlySpan<T> to implement IEnumerable<T>

We need an new interface to carry the enumerator type instead of boxing it into `IEnumerator<T>`:

```cs
interface IEnumerable<T, TEnum> where TEnum : IEnumerator<T>, allows ref struct
{
    TEnum GetEnumerator();
}

ref struct Span<T> : IEnumerable<T, SpanEnumerator<T>>
{
    SpanEnumerator<T> GetEnumerator() => new (this);
}

ref struct SpanEnumerator<T>(Span<T> span) : IEnumerator<T>
{
    public T Current { get; }
    public bool MoveNext();
}
```

But this can be hard to use without associated types, ideally it should be:

```cs
interface IEnumerable<T>
{
    type TEnum : IEnumerator<T>, allows ref struct;
    TEnum GetEnumerator();
}

ref struct Span<T> : IEnumerable<T>
{
    type TEnum = SpanEnumerator<T>;
    TEnum GetEnumerator() => new (this);
}

ref struct SpanEnumerator<T>(Span<T> span) : IEnumerator<T>
{
    public T Current { get; }
    public bool MoveNext();
}
```

**Reply by TahirAhmadov** [reply]
Unfortunately, I found an issue.
[SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAFxFANnAJiA1AHwAIBMAjALABQ5MEYApgM4AOEAxjQAQCSMUCAQigHtmAawDCAlChrMEUATAAq9BOQDe5cgEgA9NrwBmNoTYAZKHQQBBADwKAfGxCnzCW3a271ZHds0GjACxslujoABR4RPpubCguAJRsALwOCAAW5gB0IegAShAwAOY0YXEW8QDcHtoAvtX+xgCy1V4+foZmFjawCA4AGmyqbMUIFWw1yWwA2gC6Vd667c5d1gByDgCag8M0o+OTs/M+dQt6hsar2yeaPTQATlQoRgRGRADsWq1+RABsgWyNaAwCJEAAMszYEDuhTo8S0mi+Pki0R6Dh6dAORDgbAI2P0c3hSKiawcMAxSWmMBoAHcwvFsVTafS2Iy6QTvJofAA3KFsMCTRkAwmeYW+AYUoZYtjo8ZwUWaLYSlk0tirOkMjHXTm6GrzQk9CEyilTKW4tj4o7a7QAZSYMGsqLYjAKk3R2TotoKdMtPgAZgI7jQWGk2GEetLpTAnXa4RyfIidaK3QB1NL3EoAD2SDizDiI8UyCgEnQQ3sJPLusRcApVJYIDpgvTDjdhPorbF9qAQ6fQkzKCEyqfT1jrDd62NHqMyAFEYABXWh3CAIAPY1FhLMpNi5tj5n3+wPB0PhqCRjtdnuxzkIwnfACcYcqhK11xOJ1ud19LHYHFnC/uy4BvW9hqFoChsKIc53IGjbbCMYzXMAAgSACAhcjQqw0BmpZPmQ76NvcX6sJwf6LhAwBSMB7hkK0ui/vOZErncVFsAA4rspEAUxZZ4eQH5ET+nFLhRNDAdiChCYBdwOE49H/sJlFuFo1JpoGbASQxXEBo4JGaUuTFUaB3iChp8lSWxHF6VJPEnJEfzGGYMAAI7TthRlLHZbBDoGkkibYvlSOJklMeJDjeTQ/lWSJQVWSF6l2GE6TmOpAXsHQAhQaw2J4AArLY2JIRIDgMIG6BQMwy40FemgqemKVRVIOlyWRIlifVZlMdRnK1WppmMdpsnBUBSneF8eBvMq1JeapNCpZFZnRe1/V3KFYTpZlNDYiVNBlRV3a4Zob7kIe6DyCgACeToIHcc4yNN6ZzX1AGLU9+mrvFTWpW14VDXc83LaFXU9ewr3kY1g0NaJ+VLVp0nKTNMNvZWEMdcNIE0VoJ1nZdoMiWwAD6613KwRxYzAF1GHlCgFchKAOPj227ZVereP4P2Q2EuONUTWWU9DhV02wjPlZVV5fEldCZITGXE+wFI8zQloS1Lwt7XLQulSL3ZHCcWhszNv3/bDoUWQgv10p88LjZN90+bFAZG0jq3K9LG2ZOxZv23c6psC7quizrGinM1sMsSHClQ/Y7uWaj3tXiKpx+BNyse+bB2JrxZAWDdd3s7HjtSSbKPLYZ3jA4j5nF6HI2tPrD1e5zv1OjLvO5fztPFZrati/CLsK5MCtKxkksM13lWTP72taLrrOGOBkHQTQsFbn3Lc0JkC8waMQdLALqHoZh2EW6N8Iqag7CJcPUsK5kjRoRhWE4dViJQL6oZ+2P3aX1kruyxvUFb3iNVTk1trpzkVvCLU1svwoDoBA7wM9NBkwpqDJiBNB6YyDKdcml027UzYALemk94Ezwci4FiVcI5tVHFHX61FWglSgDybsbBkE4whPjKES5zosyWHWTmEIuEQHOj3DkLshGXQpBIwOs82BN1Tg3URICJqCnNhLA6iC6K/TDl9Nw0dPax2PvGaoSdfZXwUYY9OtQd7+GzrdBAcivafS9qXG8HI2HLAQCxfG/YCCWnDPjWA6AsKTAALREF4Usc2NCHC+MSE4CWRi3HXhdr4vs5DLSIKWPPABS8HEryvj48hUsJFTBdkErC7IraGD3nfA+j8kmImtlgLA5SYDBKzNYMxP9fElOgsIzIJgl6FHSJk6e5AahAA===)

I'm not sure if it can be changed for `ref struct`, but I doubt it. Here is the problem: under current rules, C# copies the struct-typed field before invoking a method on it, when that field's type is a generic parameter. This means the `MoveNext`'s state changes are lost, precluding the idea of a no-allocation enumeration.
Unless I'm missing something ...

**Reply by agocke** [reply]
@TahirAhmadov You're missing something.

You marked the enumerator as `readonly`:

> 	readonly TEnumerator _source;

That's causing the copy when you call `_source.MoveNext()`. You need to leave the field as mutable.

**Reply by TahirAhmadov** [reply]
Yes of course! Made it readonly by habit 🤣

**Reply by jnm2** [reply]
`readonly` on fields being an attractive nuisance once again :)

**Reply by TahirAhmadov** [reply]
Really though, another item in the "when we re-write everything from scratch" bucket:
Disallow emitting "copy" in IL. If a struct method is non-readonly, prohibit calling on return values, readonly fields, etc.

**Comment by TonyValenti** [comment]
For me, the ref struct implementing interfaces would be super nice for span/enumerables. That is my main use case.

Maybe this was considered, but, perhaps instead of making spans implement interfaces, there should be a method to have the compiler generate a wrapper type that implements the interface.

**Reply by KennethHoff** [reply]
> Compiler generate a wrapper type that implements the interface.

That's basically the plan for "Extensions implementing interfaces", so one could potentially utilize the same compilation strategy for that - whatever that turns out being - similar to how `params collections` utilized `collection expressions`.

**Reply by TahirAhmadov** [reply]
The wrapper type would need to be a `ref struct` in order to hold a `Span`, so in order for it to implement interfaces, we still need `ref struct` to be able to implement interfaces and at that point, why not just add it to `Span`? Or I misunderstood the point you were trying to make?

**Reply by TonyValenti** [reply]
I have never actually rolled my own ref structs and have really only used Span/ROSpan.

Here is an interesting idea:
Is this something that should be handled by C# or the runtime?

If I understood one of the challenges, it had to do with interface dispatch. 

Would a method like this allow the runtime to specialize that away?

```
Public static void Foo<TCollection, TEelement>(TCollection source) where TCollection : IEnumerable<TElement>
Where TElement allows ref struct {

}

```

Pardon typos. Typing on my phone.

**Reply by yaakov-h** [reply]
@TonyValenti it wouldn't generally help because you would still have the enumerator from `IEnumerable<TElement>.GetEnumerator()` as `IEnumerator<TElement>`. You could have your collection but you would never be able to perform useful operations over it.

**Comment by ufcpp** [comment]
I am wondering if it could be used to implement Lists with ref struct whose content is almost the same, only the Resize policy is different.
https://github.com/ufcpp/UfcppSample/tree/master/Demo/2024/ValueList/Collections

**Comment by hez2010** [comment]
Worth to note that the runtime has added APIs to allow using ref structs for comparers, so that one can implement `IComparer<T>` or `IEquatable<T>` for a ref struct. This is definitely a use case for ref struct implementing interfaces.

**Comment by neon-sunset** [comment]
I'm an author of https://github.com/U8String/U8String project (the one that uses cursed source generator api). Due to ever increasing amount of APIs, concerns regarding internal structure in combination with future extensibility and lack of time, the project has been progressing slowly.

However, upon evaluating ref structs implementing interfaces and passed as generics, it appears that this exact feature will allow to unify and generalize significant amount of internal logic, bringing the approach closer to how Rust generalizes its own string implementations.

Currently, `U8String` has a lot of logic for slicing, splitting, enumerating and otherwise manipulating UTF-8 strings. This logic, however, is hard tied to `U8String` itself. This sometimes clashes with C#'s de-facto UTF-8 string type being `ReadOnlySpan<byte>`. The implication of this is users are expected to either start with `U8String` as a data type from the beginning, which the library tries to make as convenient as possible, or to reallocate `ROS<byte>`s into new `U8String`s instead.

With this change, I have a prototype that accepts `TU8String` (or even just `TString`) that can serve both `U8String` and `U8Span` implementing the common set of interfaces, allowing cleaner integration with `ROS<byte>` and lighting it up for scenarios like splitting without duplicating certain types and methods. This could be expanded in the future to e.g. `NativeString` and others.

**Comment by MitchRazga** [comment]
Bit of a niche use case but `ref struct`s implementing interfaces would allow me to reduce the number source generator marker attributes that I use to workaround the absence of  `ref record struct`s.
```csharp
...
[WithInterfaceProperties]
public readonly ref partial struct Inputs : IAreaThresholds, IDeviationThresholds, IMeanThresholds;
```
vs
```csharp
...
[WithInterfaceProperties<IAreaThresholds>]
[WithInterfaceProperties<IDeviationThresholds>]
[WithInterfaceProperties<IMeanThresholds>]
public readonly ref partial struct Inputs;
```

**Comment by ds5678** [comment]
Thank you all so much! I refactored [my HTML generation library](https://github.com/AssetRipper/AssetRipper.Text.Html) to make heavy use of ref struct interfaces.

In my case, all the HTML element types are `ref struct`s, in order to encourage proper API use and cultivate a pit of success.

The introduction of ref struct interfaces allows me to write code more generically. One example is the [`CustomAttributes`](https://github.com/AssetRipper/AssetRipper.Text.Html/blob/3054d553836cfedb49b99dfc40ac75604cd6a1f6/AssetRipper.Text.Html/CustomAttributes.cs) class which provides support for emitting attributes that have not been pre-defined. Previously, these methods were generated into every HTML element type.

**Comment by TahirAhmadov** [comment]
In case it wasn't already brought up: imagine you have a lazy-instantiated collection, and a "proxy collection" to wrap around it:
```cs
class Foo
{
  public IList<string> Strings => this._list ??= [];
  public Wrapper Numbers => new Wrapper(ref this._list);

  List<string>? _list;
}
public readonly ref struct Wrapper : IEnumerable<int> // currently, this is CS8343
{
  public Wrapper(ref List<string> list)
  {
    this._list = list;
  }
  public readonly void Add(int i)
  {
    (this._list ??= []).Add(i.ToString());
  }

  readonly ref List<string>? _list;

  public IEnumerator<int> GetEnumerator() => throw new NotSupportedException();
  IEnumerator IEnumerable.GetEnumerator() => throw new NotSupportedException();
}

var foo = new Foo
{
  Strings = { "abc" },
  Numbers = { 123 } // currently, this is CS1918
};
Console.WriteLine(string.Join(", ", foo.Strings));
// output: abc, 123
```

To accomplish something similar, I have to do some weird stuff with delegates and such.

**Reply by TahirAhmadov** [reply]
I just noticed another issue - CS1918 is generated because it's a struct, I think it shouldn't be an error for collection initalization, or probably an even more encompassing rule - if member is readonly, allow.
I ran some memory profiling and working through excessive allocations, it would be cool to have this feature. 
@333fred does this sound reasonable? Should I open a discussion for it?

**Reply by 333fred** [reply]
Honestly, this example is very confusing. Why is that a ref struct? Where's the implementation of IEnumerable? What's the error you're getting? What are you suggesting changing?

**Reply by TahirAhmadov** [reply]
Please my edits above. It's a `ref struct` so that I can store the underlying `List<string>` as a ref field. Then the lazy instantiation can happen whether I add to `Strings` itself or to `Numbers` in an object initializer. `IEnumerable` serves solely as a marker to enable collection syntax in object initializer; it's not meant to be used by itself, therefore it just throws. It's generating CS8343, which is what this discussion is about, but also CS1918. I think it should be allowed to invoke stuff on a struct in object initializer within some boundaries, specifically, if the members are `readonly`.

**Reply by 333fred** [reply]
@TahirAhmadov for implementing IEnumerable, it sounds like you're not using a current compiler in preview mode. As to the other suggestion, you could certainly open a discussion.

**Reply by TahirAhmadov** [reply]
https://github.com/dotnet/csharplang/discussions/8427
